#435 Build Side Navigation

sandrina-p opened the pull request:
Guys, fell free to take a deep look at the new side navigation:

<img width="1198" alt="screen shot 2018-08-03 at 14 01 15" src="https://user-images.githubusercontent.com/14869087/43670534-6e6f777e-9728-11e8-8d35-c4ea0512a26f.png">

## What's missing:
- [x] Hide sidebar on "Home"
- [x] Adjust sidebar content if it's logged or not
- [x] Delete original navbar
- [x] ~~Delete top header on homepage.~~ let's keep it for now.

- [x] Develop "Show More" part
- [x] Extract "Profile" logic from "NavBar" to its own component
- [x] Move "notification" logic implemented at original NavBar - Activity.vue
- [x] MVP responsive - only a menu drawer toggle

- [x] Some misc stuff... Tests, Styles, A11Y, DRY, etc...

@taoeffect and @hubudibu there's some `QUESTION:` comments about stuff I didn't know how to do or I'm not sure about the solution. When possible check it out, thank you!

### Notes
#### Custom CSS classes prefix

So far we've been adding `.gi-` prefix to all custom CSS classes so we don't risk being overwritten by Bulma. I've been finding it confusing lately, asking the same question: _Is this a global .gi class or only used in this component?_

To make things easier, I suggest an improvement here with a new prefix `.c`:
- **`.gi-`** -> Global custom classes, ex: `gi-is-justify-between`, `gi-is-ellipsis`
- **`.c-`** -> Scoped classes used on that component only, ex: `.c-item`, `.c-icon`.

This makes things easier when reading the HTML to know where a class come from:

What do you think guys?  

### Dependencies
#### Update Vue to 2.3.4 -> 2.4.4
- Now we can  [Disabling-Attribute-Inheritance](https://vuejs.org/v2/guide/components-props.html#Disabling-Attribute-Inheritance)

#### Add package [`vue-clickaway`](https://github.com/simplesmiler/vue-clickaway)
- It was needed at `<MenuSelect />` to detect clicks outside the dropdown. For sure we will need in future cases, tooltips, modals, etc... There was another package [`vue-click-outside`](https://www.npmjs.com/package/vue-click-outside) but didn't work correctly, so I chosen  `vue-clickaway` that also has a bigger contribution graph.

sandrina-p committed:
435 - Redesign navbar from vertical to horizontal

- Refactor Navbar component: Navbar, Activiy, GroupsList and Profile
- Create Menu component
- Create Item component

taoeffect commented on frontend/simple/views/components/Lists/index.js:
```diff
@@ -0,0 +1,2 @@
+export { default as List } from './List.vue'
+export { default as ListItem } from './ListItem.vue'
```
Would it be possible to avoid these `index.js` files?

taoeffect replied:
I mean, it feels like there might be a cleaner way of doing this... However, I see that you at least do fully specify the filename and extension when importing, so it's not too bad... just wondering if there's some other way. I'll keep thinking on it as I review the PR in more detail.

sandrina-p replied:
This is the cleaner way I know to export multiple named modules. Maybe @hubudibu knows something better?

hubudibu replied:
I also prefer these index.js files, it helps keep the imports where these components are used shorter. it makes it possible to do
```js
import { List, ListItem } from '../../components/Lists'
```
instead of having to do longer imports for each line like
```js
import List from '../../components/Lists/List.vue'
import ListItem from '../../components/Lists/ListItem.vue'
```
it also makes it more explicit at the imports where we use these that they go together.

taoeffect replied:
Yes but per the recent discussions all extensions must be specified, so we can't do `import { List, ListItem } from '../../components/Lists'`, it must be `import { List, ListItem } from '../../components/Lists/index.js'` for future compatibility and also clarity. But it's fine to do it that way. 👍

sandrina-p replied:
Kidoki, I'll add the `index.js`!

sandrina-p replied:
Added

taoeffect commented on .Gruntfile.babel.js:
```diff
@@ -111,7 +111,7 @@ module.exports = (grunt) => {
         options: {env: {LOAD_NO_FILE: 'true', ...process.env}}
       },
-      standard: './node_modules/.bin/standard "**/*.{js,vue}"',
+      standard: './node_modules/.bin/standard "**/*.{js,vue}" --fix',
```
I feel like this isn't the first time I'm making this comment about this same addition, but to be clear — `--fix` **must** be run manually. It cannot be run automatically every single time a file changes.

I hope the security and bug-related implications of that are clear and I do not need to spell them out. It is already a security risk that we're using `standard` at all.

sandrina-p replied:
I perfectly understand and I'm aware of it. I'm sorry, I did stage the file without noticing it, I use it this way only locally. I undid the change.

taoeffect commented:
```
To make things easier, I suggest an improvement here with a new prefix .c:

.gi- -> Global custom classes, ex: gi-is-justify-between, gi-is-ellipsis
.c- -> Scoped classes used on that component only, ex: .c-item, .c-icon.
This makes things easier when reading the HTML to know where a class come from:

What do you think guys? 
```

I honestly do not like either of these prefixes. We have two open issues for getting rid of Bulma (#392 and #372), so I do not see much of a reason for using them.

If you really want to use them, I would suggest sticking with `.gi`, because at least the meaning of that is self-explanatory. `.c` is confusing and will make people wonder what it is. Your explanation above will not be accessible to future developers, and it will have to be explained to everyone who encounters the source what it means.

So, in summary, I am for getting rid of both of them, and at most keeping only `.gi`.

EDIT: also, if you're concerned about conflicts of classes between vue components, you can always use [Vue.js's `scoped` feature](https://vue-loader.vuejs.org/guide/scoped-css.html).

taoeffect commented on frontend/simple/assets/sass/_groupincome.scss:
```diff
@@ -16,6 +16,36 @@
+// ------ index.html ------ //
+// QUESTION: This doesnt look like the best place
+// to style the index.html / main.js
+// Why isn't that file a vue file like all the others?
```
`index.html` and `main.js` cannot be `.vue` files because `.vue` files exist solely for the purpose of getting compiled into `.js` file(s) that are then loaded via the main js bundle.

In other words, `index.html` and `main.js` are bootstrapping files for the whole system, and `.vue` files are components that get compiled into a javascript bundle.

Basically, it's just because of the way that JavaScript, HTML, and VueJS work that it's this way.

sandrina-p replied:
Hum... I see... Ok

taoeffect commented on frontend/simple/main.js:
```diff
@@ -41,7 +41,15 @@ async function loadLastUser () {
+        console.log('router:::', router.name)
+
+        // QUESTION: How to I hide the sidebar if the router is "home" ?
+        return true // router.name === 'home' // undefined...
```
I think it would be best to let me and/or @hubudibu think about how to handle this part. For this PR if you could delete this it and open an issue and assign it to one or both of us it would be helpful.

sandrina-p replied:
Done. https://github.com/okTurtles/group-income-simple/issues/437

sandrina-p replied:
Actually I find it better the nav to be hidden until the user is logged. `this.$store.state.loggedIn`

taoeffect commented on frontend/simple/views/components/Lists/ListItem.vue:
```diff
@@ -0,0 +1,146 @@
+
+      For now IMO "Duplication is better than the wrong abstraction"
+      until we found a solid solution for this problem
+    -->
```
OK, for this can you create an issue and assign to @hubudibu?

sandrina-p replied:
I've made this question at stackoverflow a few days ago and someone helped me! [Conditional Parent Element on Vue.js 2](https://stackoverflow.com/questions/51293489/conditional-parent-element-attrs-on-vue-js-2)

taoeffect commented on frontend/simple/views/components/Menu/MenuItem.vue:
```diff
@@ -0,0 +1,26 @@
+    and to update MenuSelect when an option was selected.
+
+    QUESTION: Is v-bind="$attrs" the correct way to pass down
+    the props and slot from MenuItem to ListItem?
```
I don't know off-hand, I see this is [mentioned in the docs](https://vuejs.org/v2/api/#vm-attrs), but I would need to spend a lot more time researching this to be able to give you a good answer. So while my attention is tied up elsewhere perhaps @hubudibu might be able to help here?

sandrina-p replied:
Yes, exactly, so I felt comfortable in doing it this way.

hubudibu replied:
yes, it's a totally correct way. I would however prefer to unbundle these attrs and pass them down one-by-one, to make it possible to do type-checks on them, and to make it more explicit reading the code what props this component needs. but that's probably unnecessary at a simple wrapper component like this. it's the usual compactness vs self-documenting readability question :)

sandrina-p replied:
This is exactly to give totally freedom to the developer to pass whatever props (s)he needs at the moment, it's almost like and HOC (in react vocabulary)

taoeffect commented on package.json:
```diff
@@ -140,6 +147,7 @@
     "stylelint-config-standard": "18.2.0",
     "through2": "2.0.3",
     "vue-hot-reload-api": "2.1.0",
+    "vue-template-compiler": "2.4.4",
```
Why has this been added?

sandrina-p replied:
It was a wrong install. I was suppose to update `vueify`, that has `vue-template-compiler` as dep, to match with `vue` version. It's fixed now.

sandrina-p commented on frontend/simple/views/containers/NavBar/NavBar.vue:
```diff
@@ -0,0 +1,86 @@
+
+  /* Gradient or solid bg */
+  background: linear-gradient(210deg, rgba($primary, 0.15), $body-background-color 20rem);
+  background: $primary-bg-s;
```
@taoeffect here is where you can test both backgrounds to the navbar.

sandrina-p commented:
@taoeffect regarding class prefix. We need to use a prefix because, for example `.c-header`, that class without prefixes exists on Bulma `.header` and it will apply styles that aren't needed. Using a prefix solves the problem. 

And Vue.js's scoped feature does not fix the problem of conflicts totally because it's scoped to the DOM element, not to the file where it's created (like CSS modules or CSS-in-JS).

sandrina-p commented on test/frontend.js:
```diff
@@ -24,7 +24,6 @@ function login (username) {
     n
       // Logout
-      .click(elT('openProfileDropDown'))
       .click(elT('logoutBtn'))
```
@taoeffect and @hubudibu I have a question. The test failing is
> 1) Frontend Sign up Test Test Logout and Login:
> Error: Timeout of 10000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves.

I suspect the problem is inside this login function.

`'openProfileDropDown'` doesn't exist anymore, now it's only `'logoutBtn'`, so I've removed that line.

But the test is still failing. Can you explain me better why the first step of `login()` is to Logout? Can I better debug these tests? Do breakpoints and see the browser at that exact moment, etc...

taoeffect replied:
You can run the tests and see them visually:

```
$ SHOW_BROWSER=1 grunt test
```

And you can use Mocha's `.only` and `.skip` modifiers to only run certain tests and skip others (see the Mocha documentation).

You can also insert purposeful `.wait(1000)`s to slow down the tests.

taoeffect commented:
@sandrina-p I noticed that the `package-lock.json` file in this PR is really weird and does weird things. Now that my recent PR was merged, there's a conflict with it, but also I suspect that the one that was generated by you might have been generated by an outdated version of npm.

I really hate that file because of how difficult it is to fix these kinds of issues.

Here's what ideally should happen in order to restore/fix the integrity of that file:

- A branch off of the current `master` should be created
- All changes in this PR **except for the changes to `package.json` and `package-lock.json`** should be applied to it manually somehow. Then, making sure **first** that you have at least node version 9.x (preferably 10.x), and npm version 6.2 or greater, run `npm i vue@latest vue-clickaway -S`
- This will result in a single commit that you can then re-push in a different PR (closing this one)

sandrina-p committed:
Reset package-lock.json with master and install vue-clickaway

sandrina-p committed:
just force another travis pipeline

hubudibu commented:
I had a similar issue with Travis not finding paths when I deleted a `something.vue` file and created a `something` folder instead. Maybe this is what's happening here as well with NavBar.vue and the NavBar folder. You could rename the NavBar folder anyways, as it's not a bar anymore, and maybe it will fix our Travis troubles too?

hubudibu commented on test/frontend.js:
```diff
@@ -130,14 +128,13 @@ describe('Frontend', function () {
   describe('Sign up Test', function () {
     it('Should register User', function () {
-      this.timeout(10000)
+      this.timeout(100000)
```
that's a lot :)

sandrina-p replied:
Fixed

hubudibu commented on test/frontend.js:
```diff
@@ -327,7 +324,12 @@ describe('Frontend', function () {
 
       const alert = await n.exists(elT('alertNotification'))
+
+      console.log('alert:::::::', alert)
```
now you can use the amazing new `note` feature for debugging stuff like this

sandrina-p replied:
Yes, this was before this new era, and I forgot to delete it

hubudibu reviewed (commented):
super happy to see the sideeffect of all the component HTMLs getting shorter and cleaner with every PR @sandrina-p makes ▶️◀️

sandrina-p committed:
Update code based on review

- Rename Navbar to Sidebar - fix travis finding paths issue, I hope
- Remove navbar bulma css dependecies (more than 400 lines of CSS)
- Misc: Delete some old lost console.log on tests and fix indentations

hubudibu commented:
and about the `.c-` class prefix: I see why it's needed - the difference between global custom styles (gi-) and component level custom styles (c-), and as @sandrina-p pointed out, this is a widely accepted practice with UI libraries, so it should be well known among UI developers and can be picked up without much explanation 👍

sandrina-p committed:
Rename Sidebar/ to sidebar/

sandrina-p committed:
Adjust some PixelPerfect Miscs.

- Udpdate $grey color
- Adjust badge size
- Set sidebar bg to solid light blue
- Fix bug on small screens when sidebar content is bigger than the screen

sandrina-p commented:
![image](https://user-images.githubusercontent.com/14869087/43689444-8ff8ba12-9896-11e8-9c00-1a7ac86c9657.png)

Guys, I've updated (once again) the sidebar. Now it has a solid light blue background and some greys were improved for better readability. Please let me know what you think.

taoeffect commented on frontend/simple/views/components/Menu/MenuParent.vue:
```diff
@@ -0,0 +1,56 @@
+    },
+    handleSelect (itemId) {
+      this.closeMenu()
+      this.$emit('select', itemId)
```
- What are these two events for ('open' and 'select')?
- What is listening to them?
- Should they use SBP events instead?

sandrina-p replied:
Components that use `<menu-parent>` can listen for those events.

ATM `GroupsList.vue` is listening for `select` and once `Activity.vue` works correctly it will do it too (on a next PR).

I didn't use SBP here because `this.$emit` / `@select` ([Sending Custom Events to Parents](https://vuejs.org/v2/guide/components.html#Sending-Messages-to-Parents-with-Events) is a specific Vue feature.

```
<menu-parent @select="handleMenuSelect">
```

It would be the same as doing the following:


```js
handleSelect (itemId) {
      this.closeMenu()
      this.onSelect(itemId) // onSelect being a prop
}
```

```vue
<menu-parent onSelect="handleMenuSelect">
```

but I found `@` prop more readable. Also, so far, sbp looks like it's global, and here only the parent can listen for these events.

ATM `open` prop isn't being listen anywhere, so I'll delete it and once is needed we can add it again.

taoeffect commented on frontend/simple/views/containers/SignUp.vue:
```diff
@@ -1,3 +1,4 @@
+<!-- TODO: Transform this view into a Modal, for layout consistence -->
```
No longer doing this so this comment can be deleted

sandrina-p replied:
Indeed. removed.

taoeffect commented on frontend/simple/views/containers/sidebar/Activity.vue:
```diff
@@ -0,0 +1,69 @@
+        <menu-header>
+          <i18n>Notifications</i18n>
+        </menu-header>
+        <list hasMargin>
```
How does this component receive / know about `hasMargin`?

taoeffect replied:
Oh, nevermind, it is creating a list and is telling it that it has the margin... I see.

taoeffect replied:
I'm not sure I like this. This is confusing. Why do it this vue-specific way when normal CSS classes work just fine?

taoeffect replied:
General rule of thumb: if there are two different ways to do something, and both do the same thing, and one is simpler than the other, do it the simpler way.

Using this vue.js feature when it's not needed and all it does is apply a class is using CSS + Vue.js feature, and unclear why the latter is necessary.

It would work just the same with just pure CSS, right? Then that is the simpler and more correct solution.

sandrina-p replied:
When tomorrow we have 3 places with the same css rule, we'll have `margin: 1rem 0` duplicated 3 times. So this prop prevents that from happening.

_Still..._, what I've done is what it's know as _wrong abstraction_ predicting the future. And human beings are horrible predicting the future. So, I'll leave it as raw css and when the duplication comes in (if it does) we abstract it.
Until then, _Prefer duplication over the wrong abstraction_

sandrina-p replied:
Actually, we already have 2 places where this `prop` is being used (and not only one as I thought), so I think the prop is okay here.
If we use only a class, we would have to write

```scss
.gi-list {
  margin: $gi-spacer-sm 0;
}
```

twice in our code... (in this case `Activity.vue` and `GroupsList.vue`). With a prop we avoid that.

taoeffect replied:
Why? Just put it into the global scss?

sandrina-p replied:
IMO I don't consider this a global class, when it's specific to a component only.

And I would like to avoid adding global classes all the time, otherwise in a near future our "global CSS scope" would be full of trash and probably lost classes.

sandrina-p committed:
Remove stale comments and delete not used prop @open at Menu component

sandrina-p committed:
Add comments to explain some props

sandrina-p committed:
Fix Join.vue width on desktop - use reusable class

sandrina-p committed:
Update Activity.vue with Lists import by adding
